carma-messenger-core/cpp_message/src/Map_Message.cpp

```diff
@@ -0,0 +1,614 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+#include "Map_Message.h"
+
+namespace cpp_message
+{
+    boost::optional<j2735_msgs::MapData> Map_Message::decode_map_message(std::vector<uint8_t>& binary_array)
+    {
+        j2735_msgs::MapData output;
+
+        // decode results
+        asn_dec_rval_t rval;
+        MessageFrame_t* message = nullptr;
+        
+         //copy from vector to array         
+        size_t len=binary_array.size();    
+        
+        uint8_t buf[len];             
+        std::copy(binary_array.begin(),binary_array.end(),buf);
+        //use asn1c lib to decode
+        
+        rval=uper_decode(0, &asn_DEF_MessageFrame,(void **) &message, buf, len, 0, 0);
+        if(rval.code == RC_OK)
+        {
+
+
+            auto map_msg = message->value.choice.MapData;
+
+            IntersectionGeometry_t *map_msg_intersections = new IntersectionGeometry_t;
+
+            //Intersection Handling
+            if(map_msg.intersections)
+             {   
+
+                output.intersections_exists = true;
+                //Map Intersections
+                for(size_t i = 0; i < map_msg.intersections->list.count; i++)
+                {
+                    map_msg_intersections = message->value.choice.MapData.intersections->list.array[i];
+                    j2735_msgs::IntersectionGeometry new_intersection;
+                    new_intersection.id.id = map_msg_intersections->id.id;
+                    new_intersection.id.region_exists = true;
+
+                    //Lane Set 
+                    for(size_t l = 0; l < 0; l++)
+                    {
+
+                        GenericLane_t *lane = new GenericLane_t;
+
+                        lane = map_msg_intersections->laneSet.list.array[l];
+                        j2735_msgs::GenericLane ln;
+
+                        if(lane->egressApproach)
+                        {
+                            ln.egress_approach_exists = true;
+                            ln.egress_approach = *lane->egressApproach;
+                        }
+                        else
+                        {
+                            ln.egress_approach_exists = false;
+                        }
+
+                        if(lane->ingressApproach)
+                        {
+                            ln.ingress_approach_exists = true;
+                            ln.ingress_approach = *lane->ingressApproach;
+                        }
+                        else
+                        {
+                            ln.ingress_approach_exists = false;
+                        }
+
+                        if(lane->maneuvers)
+                        {
+                            ln.maneuvers_exists = true;
+                            ln.maneuvers.allowed_maneuvers = *lane->maneuvers->buf;
```

Comment:
Here we are trying to convert the bit string to uint16, this would not be the correct way to do this.
We need to go through the bit array and assign values based on that. There are examples of this in the other message conversions.